### PR TITLE
feat: Implement a basical range select query

### PR DESCRIPTION
**Pull request by Taylor-lagrange**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Implements the physical plan portion of the range select query

Please explain IN DETAIL what the changes are in this PR and why they are needed:

The execution process of range select query is relatively straightforward. It mainly refers to the row hash aggregation of datafusion.  

For each `align_ts` time on each time series we use a `Accumulator` to record data to be aggregated

The main execute process are:
1. receive a `RecordBatch` from upstream.
2. caculate the hash value of by columns
3. collect the row nums affected for each  `Accumulator` 
4. update data in `Accumulator` 
5. Evaluate each `Accumulator` until the end of input and return the result

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)

https://github.com/GreptimeTeam/greptimedb/issues/1662

### CODE REVIEW

**Comment by killme2008** [comment]
@Taylor-lagrange can you rebase this branch with the latest develop branch? Thank u.

File: src/query/src/range_select/plan.rs [hunk 1]
```diff
@@ -216,30 +358,226 @@ impl ExecutionPlan for RangeSelectExec {
...
-pub struct RangeSelectStream {
+struct RangeSelectStream {
+    /// the schema of output column
     schema: SchemaRef,
+    range_exec: Vec<RangeFnExec>,
     input: SendableRecordBatchStream,
+    /// Column index of TIME INDEX column's position in schema
+    time_index: usize,
+    /// the unit of `align` is millisecond
+    align: i64,
```
Comment (waynexia) [inline src/query/src/range_select/plan.rs:424]: Millisecond has its own type alias

File: src/query/src/range_select/plan.rs [hunk 2]
```diff
@@ -216,30 +358,226 @@ impl ExecutionPlan for RangeSelectExec {
...
     }
 }
 
-pub struct RangeSelectStream {
+struct RangeSelectStream {
+    /// the schema of output column
     schema: SchemaRef,
+    range_exec: Vec<RangeFnExec>,
     input: SendableRecordBatchStream,
+    /// Column index of TIME INDEX column's position in schema
```
Comment (waynexia) [inline src/query/src/range_select/plan.rs:421]:
```suggestion
    /// Column index of TIME INDEX column's position in the input schema
```

File: src/query/src/range_select/plan.rs [hunk 3]
```diff
@@ -216,30 +358,226 @@ impl ExecutionPlan for RangeSelectExec {
...
+            .unwrap();
+        for i in 0..self.range_exec.len() {
+            let args = self.evaluate_many(&batch, &self.range_exec[i].args)?;
+            // use self.modify_map record (hash, align_ts) => [row_nums]
+            self.modify_map.clear();
+            let range = self.range_exec[i].range;
+            // make modify_map for range_fn[i]
+            for (row, hash) in hashes.iter().enumerate() {
+                let ts = ts_column.value(row);
+                let mut align_ts = ((ts + self.align - 1) / self.align) * self.align;
```
Comment (waynexia) [inline src/query/src/range_select/plan.rs:485]: Define this as a method. We need to implement other align methods in the future (align to calendar or execute time)

File: src/query/src/range_select/plan.rs [hunk 4]
```diff
@@ -252,12 +590,230 @@ impl Stream for RangeSelectStream {
...
+        let timestamp_column = Arc::new(TimestampMillisecondArray::from(vec![
+            // host 1 every 5s
+            0, 5_000, 10_000, 15_000, 20_000, 25_000, 30_000, 35_000, 40_000,
+            // host 2 every 5s
+            0, 5_000, 10_000, 15_000, 20_000, 25_000, 30_000, 35_000, 40_000,
+        ])) as _;
+        let values = vec![
+            // host 1
+            0, 1, 2, 3, 4, 5, 6, 7, 8, // host 2
+            9, 10, 11, 12, 13, 14, 15, 16, 17,
```
Comment (waynexia) [inline src/query/src/range_select/plan.rs:662]:
```suggestion
            // host 1
            0, 1, 2, 3, 4, 5, 6, 7, 8,
            // host 2
            9, 10, 11, 12, 13, 14, 15, 16, 17,
```

File: tests/cases/standalone/common/select/range_select.result
Comment (waynexia) [inline tests/cases/standalone/common/select/range_select.result:1]: Also add some negative (invalid) cases

File: src/query/src/range_select/plan.rs [hunk 5]
```diff
@@ -80,7 +99,10 @@ impl RangeSelect {
                 Ok(DFField::new_unqualified(
                     &expr.display_name()?,
                     expr.get_type(input.schema())?,
-                    expr.nullable(input.schema())?,
+                    //TODO: We have not implemented fill currently,
```
[the remark this answers is not in the document]
Comment (killme2008) [reply src/query/src/range_select/plan.rs:102]: The format of `TODO` must be `TODO(github username)`

File: src/query/src/range_select/plan.rs [hunk 6]
```diff
@@ -216,30 +358,226 @@ impl ExecutionPlan for RangeSelectExec {
...
+    /// value: [row_ids]
+    /// It is used to record the data that needs to be aggregated in each time slot during the data update process
+    modify_map: HashMap<(u64, i64), Vec<u32>>,
+    /// The number of rows of the final output
+    output_num_rows: usize,
+    metric: BaselineMetrics,
+}
+
+struct SeriesState {
+    /// by values write by `RowWriter`
```
Comment (killme2008) [inline src/query/src/range_select/plan.rs:443]:
```suggestion
    /// by values written by `RowWriter`
```

File: src/query/src/range_select/plan.rs [hunk 7]
```diff
@@ -216,30 +358,226 @@ impl ExecutionPlan for RangeSelectExec {
...
+        let num_rows = batch.num_rows();
+        let by_arrays = self.evaluate_many(&batch, &self.by)?;
+        let mut hashes = vec![0; num_rows];
+        create_hashes(&by_arrays, &self.random_state, &mut hashes)?;
+        let by_rows = self.row_converter.convert_columns(&by_arrays)?;
+        let ts_column = batch
+            .column(self.time_index)
+            .as_any()
+            .downcast_ref::<TimestampMillisecondArray>()
+            .unwrap();
```
Comment (killme2008) [inline src/query/src/range_select/plan.rs:476]: Return an error instead of panic.

**Review by killme2008 — COMMENTED** [review]
Great work! I have one question:
* Looks like the feature only supports TimestampMillisecond type, can we support other timestamp types?

File: src/query/src/range_select/plan.rs (hunk 7 above)
Comment (Taylor-lagrange) [reply src/query/src/range_select/plan.rs:476]: Here follow the implementation of PromQL physical plan like `src/promql/src/extension_plan/instant_manipulate.rs`, I checked times index data type in logical plan stage,  the cast can not fail

**Comment by Taylor-lagrange** [comment]
> Looks like the feature only supports TimestampMillisecond type, can we support other timestamp types?

The implementation mostly refer to PromQL implemtation, the PromQL  seem only support `TimestampMillisecond type`, so I follow the patern @killme2008

**Comment by waynexia** [comment]
> > Looks like the feature only supports TimestampMillisecond type, can we support other timestamp types?
> 
> The implementation mostly refer to PromQL implemtation, the PromQL seem only support `TimestampMillisecond type`, so I follow the patern @killme2008

Here supporting other resolutions are viable. Unlike Prometheus' protocol have restrictions on timestamp type, this is a pure SQL interface functionality.

File: src/query/src/range_select/plan.rs [hunk 8]
```diff
@@ -92,17 +114,20 @@ impl RangeSelect {
...
             .context(DataFusionSnafu)?;
+        let by_schema = DFSchema::new_with_metadata(by_fields, input.schema().metadata().clone())
+            .context(DataFusionSnafu)?;
         Ok(Self {
             input,
             range_expr,
             align,
             time_index: time_index_name,
             schema: Arc::new(schema),
+            by_schema: Arc::new(by_schema),
```
[the remark this answers is not in the document]
Comment (Taylor-lagrange) [reply src/query/src/range_select/plan.rs:130]: We need provide by schema to `RowConvert` which covert by columns to row format, in order to provider by values in final output

**Comment by Taylor-lagrange** [comment]
> Here supporting other resolutions are viable. Unlike Prometheus' protocol have restrictions on timestamp type, this is a pure SQL interface functionality.

New commit has support any timestamp datatype, the different timestamp will be cast to MillSecond when range plan execute, and convert to original timestamp data type when output. 

Is more reasonable cast to  `Nanosecond` to prevent possible loss of precision？ But `u64`  is very easy to overflow if using nanoseconds

**Comment by waynexia** [comment]
>Is more reasonable cast to Nanosecond to prevent possible loss of precision？ But u64 is very easy to overflow if using nanoseconds

I think millisecond is fine 👍

**Review by killme2008 — APPROVED** [review]
LGTM

**Review by waynexia — APPROVED** [review]
Great 🚀

**Pull request merged**